src/bst/core.clj

```diff
@@ -31,99 +31,50 @@
-  [tree]
-  (if (empty? tree) false
-      (and (is-right-case? tree) (< (factor (:right tree)) 0))))
-
-(defn is-right-left-case?
-  "Returns true if right sub tree of left child is imbalanced else false"
-  [tree]
-  (if (empty? tree) false
-      (and (is-right-case? tree) (> (factor (:right tree)) 0))))
-
-(defn rotate-left
-  "Returns the left rotated tree "
-  [{:keys [root left right] :as tree}]
-  (let [pivot (:root right)
-        left-pivot (:left right)
-        right-pivot (:right right)]
-    (if (empty? tree) {}
-        {:root pivot
-         :left {:root root
-                :left left
-                :right left-pivot}
-         :right right-pivot})))
-
-(defn rotate-right
-  "Returns the right rotated tree"
-  [{:keys [root left right] :as tree}]
-  (let [pivot (:root left)
-        left-pivot (:left left)
-        right-pivot (:right left)]
-    (if (empty? tree) {}
-        {:root pivot
-         :left left-pivot
-         :right {:root root
-                 :left right-pivot
-                 :right right}})))
-
-(defn balance
-  "Returns a balanced bst"
-  [{:keys [left right] :as tree}]
-  (cond
-    (is-right-left-case? tree) (rotate-left
-                                (assoc tree :right (rotate-right right)))
-
-    (is-left-right-case? tree) (rotate-right
-                                (assoc tree :left (rotate-left left)))
-
-    (is-right-right-case? tree) (rotate-left tree)
-
-    (is-left-left-case? tree) (rotate-right tree)
-
-    :else tree))
-
 (defn insert-node
   "Returns a bst after inserting a new node"
   [{:keys [root] :as tree} value]
   (cond
     (nil? root) {:root value :left nil :right nil}
-    (neg? (compare value root)) (balance
-                                 (update tree :left insert-node value))
-    (pos? (compare value root)) (balance
-                                 (update tree :right insert-node value))
+    (neg? (compare value root)) (update tree :left insert-node value)
+    (pos? (compare value root)) (update tree :right insert-node value)
     :else tree))
+(declare create)
+(defn inorder-traversal
+  [{:keys [root left right] :as tree}]
+  (cond (empty? tree) []
+        (some? tree)  (concat (inorder-traversal left)
+                              [root]
+                              (inorder-traversal right))))
+
+(defn traverse-and-balance
+  "Returns a balanced bst given a sorted list of node values"
+  ([nodes start  end]
+   (if (empty? nodes)
+     {}
+     (let [mid (int (/ (+ start end) 2))]
+       (when (<= start end)
+         {:root (nth nodes (int (/ (+ start end) 2)))
+          :left (traverse-and-balance nodes start (- mid 1))
+          :right (traverse-and-balance nodes (+ mid 1) end)})))))
+
+(defn balance
+  "Returns a balanced bst given an imabalanced bst"
+  [tree]
+  (cond
+    (empty? tree) {}
+    (= 1 (count tree)) tree
+    :else
+    (let [nodes (inorder-traversal tree)]
+      (traverse-and-balance nodes 0 (- (count nodes) 1)))))
```

Comment:
If you are already passing in `nodes`, do you really need to pass in the first and last indices?